File: csharp-dotnet/src/Architecture.Shell.Cqrs/IRequestHandler.cs
namespace Architecture.Shell.Cqrs;

/// <summary>
/// Base handler interface for processing application-layer requests.
/// </summary>
/// <typeparam name="TRequest">The request type (command, query, notification).</typeparam>
/// <typeparam name="TResponse">The response type returned after processing.</typeparam>
/// <remarks>
/// <para><strong>Handler Requirements:</strong></para>
/// <list type="bullet">
/// <item>Each request type MUST map to exactly one handler (enforced at DI registration)</item>
/// <item>Handlers MUST respect CancellationToken and terminate early when requested</item>
/// <item>Handlers MUST use Result&lt;T&gt; for business errors, throw exceptions for infrastructure errors</item>
/// <item>Handlers SHOULD be stateless (dependencies injected via constructor)</item>
/// </list>
/// </remarks>
public interface IRequestHandler<in TRequest, TResponse>
    where TRequest : IBaseRequest
{
    /// <summary>
    /// Handles the request asynchronously.
    /// </summary>
    /// <param name="request">The request to process.</param>
    /// <param name="cancellationToken">
    /// Cancellation token for early termination. Handlers MUST check this token
    /// during long-running operations and throw <see cref="OperationCanceledException"/>
    /// when cancellation is requested.
    /// </param>
    /// <returns>The response after processing the request.</returns>
    /// <exception cref="OperationCanceledException">When cancellation is requested.</exception>
    Task<TResponse> HandleAsync(TRequest request, CancellationToken cancellationToken);
}

File: typescript-nodejs/architecture-shell-cqrs/src/IRequestHandler.ts
import { BaseRequest } from './BaseRequest';

/**
 * IRequestHandler<TRequest, TResponse> - Base handler interface for processing requests
 *
 * Handlers are single-responsibility components that process one request type.
 * Each request type maps to exactly one handler (enforced at DI registration).
 *
 * @typeParam TRequest - The request type (must extend BaseRequest)
 * @typeParam TResponse - The response type
 *
 * @remarks
 * Constraints:
 * - Each request type maps to exactly one handler (enforced at DI registration)
 * - Handlers MUST respect AbortSignal and terminate early when requested
 * - Handlers MUST use Result<T> for business errors, throw exceptions for infrastructure errors
 *
 * @example
 * ```typescript
 * export class CreateOrderHandler implements IRequestHandler<CreateOrderCommand, Result<string>> {
 *   async handle(
 *     request: CreateOrderCommand,
 *     signal: AbortSignal
 *   ): Promise<Result<string>> {
 *     // Business logic here
 *     return Result.success(orderId);
 *   }
 * }
 * ```
 *
 * @see {@link ICommandHandler} - For void command handlers
 * @see {@link IQueryHandler} - For query handlers
 */
export interface IRequestHandler<
  TRequest extends BaseRequest<TResponse>,
  TResponse = void
> {
  /**
   * Handles the request and returns a response
   *
   * @param request - The request to handle
   * @param signal - Cancellation signal for early termination
   * @returns Promise resolving to the response
   * @throws Infrastructure errors (DB failures, network timeouts)
   */
  handle(request: TRequest, signal: AbortSignal): Promise<TResponse>;
}
